blockchain/validators/src/metadata_key/validator.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Block metadata key validator.
+
+use super::{KeyValidityMap, ValidationError};
+use mc_blockchain_types::{BlockIndex, BlockMetadata};
+
+/// A validator for block metadata signing keys and signatures.
+#[derive(Clone, Debug)]
+pub struct Validator {
+    config: KeyValidityMap,
+}
+
+impl Validator {
+    /// Instantiate a validator with the given node key validity ranges.
+    pub fn new(config: KeyValidityMap) -> Self {
+        Self { config }
+    }
+
+    /// Check whether the given block metadata is valid at the given index.
+    pub fn validate(
+        &self,
+        metadata: &BlockMetadata,
+        block_index: BlockIndex,
+    ) -> Result<(), ValidationError> {
+        let ranges = self
+            .config
+            .get(metadata.node_key())
+            .ok_or(ValidationError::UnknownPubKey)?;
```

Comment:
What happens in the missing data case? Definitely for AVRs there is missing data where a key isn't existent in the records and I'm guessing there might be this case for historical data or in some odd case in the future.

There should be a case here for handling the situation where there are no historical records. Either that or we should check if it's okay to interpolate them.

The exception is if there's no missing data in these records.

Reply:
It seems easier to enumerate the SCP message signing keys (than historical AVRs) since they're configured by the node operators, and we don't have anybody generating metadata yet. This PR is in part a key revocation mechanism.

To answer your immediate question: if we're missing a key, we won't be able to validate metadata signed by that key.

Reply:
I suppose the larger philisophical question is, there's always likely to be a window of time when a node might not be participating in consensus, and so how do we indicate that gap and what do we do about it? 

In terms of representation bit:
I think in this case, indicating what ranges the message signing key is valid for works well.

In terms of action:
For non-consensus nodes syncing from a particular consensus peer or proxy representing that peer - what happens when the metadata can't be validated because that node didn't form blocks in that range? That's a subtle yet import question with a few potential answers. 

I think for unknown keys/sigs that can't be validated/keys at invalid ranges, the answer is "validator throws hard error". In the case that the data is simply missing due to a node not having formed any blocks in those ranges the answer is either record "None" as a valid answer in the client node's local block ledger or try to find a peer that does have metadata records for that range.

Reply:
Got it, I think. Yeah, we still have work to do around catch-up mode, so I just removed the check that requires block metadata from #2052 until we have a good answer there.

If we want to prohibit a node's key from appearing when it wasn't participating in consensus, we can split the range that includes this disconnect into two, one leading up to the disconnect and one after reconnecting.